src/doc/trpl/concurrency.md

```diff
@@ -0,0 +1,343 @@
+% Concurrency
+
+Concurrency and parallelism are incredibly important topics in computer
+science, and are also a hot topic in industry today. Computers are gaining more
+and more cores, yet many programmers aren't prepared to fully utilize them.
+
+Rust's memory safety features also apply to its concurrency story too. Even
+concurrent Rust programs must be memory safe. Rust's type system is up to the
+task, and gives you powerful ways to reason about concurrent code at compile
+time.
+
+Before we talk about the concurrency features that come with Rust, it's important
+to understand something: Rust is low-level enough that all of this is provided
+by the standard library, not by the language. This means that if you don't like
+some aspect of the way Rust handles concurrency, you can implement an alternative
+way of doing things. [mio](https://github.com/carllerche/mio) is a real-world
+example of this principle in action.
+
+## Background: `Send` and `Sync`
+
+Concurrency is difficult to reason about. In Rust, we have a strong, static type
+system to help us reason about our code. As such, Rust gives us two types to help
+us make sense of code that can possibly be concurrent.
+
+### `Send`
+
+The first type we're going to talk about is
+[`Send`](std/marker/trait.Send.html). When a type `T` implements `Send`, it indicates
+to the compiler that something of this type is able to have ownership transferred
+safely between threads.
+
+This is important to enforce certain restrictions. For example, if we have a
+channel connecting two threads, we would want to be able to send some data
+down the channel and to the other thread. Therefore, we'd implement `Send` on
+that type.
+
+In the opposite way, if we were wrapping a library with FFI that isn't
+threadsafe, we wouldn't want to implement `Send`, and so the compiler will help
+us enforce that it can't leave the current thread.
+
+### `Sync`
+
+The second of these two types is called [`Sync`](std/marker/trait.Sync.html).
+When a type `T` implements `Sync`, it indicates to the compiler that something
+of this type has no possibility of introducing memory unsafety when passing
+`&T`s between threads.
+
+For example, sharing immutable data with a reference count is threadsafe.
+Therefore, this type would implement `Sync`, so that it could be safely
```

Comment:
It's not clear which type you mean here -- and remember that non-atomic ref counting (as in `Rc`) is *not* threadsafe!